src/query/functions/temporal/linear_regression.go

```diff
@@ -0,0 +1,155 @@
+	"time"
+
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/ts"
+)
+
+const (
+	// PredictLinearType predicts the value of time series t seconds from now, based on the input series, using simple linear regression.
+	// PredictLinearType should only be used with gauges.
+	PredictLinearType = "predict_linear"
+
+	// DerivType calculates the per-second derivative of the time series, using simple linear regression.
+	// DerivType should only be used with gauges.
+	DerivType = "deriv"
+)
+
+type linearRegressionProcessor struct {
+	duration float64
+}
+
+func (l linearRegressionProcessor) Init(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
+	return &linearRegressionNode{
+		op:         op,
+		controller: controller,
+		timeSpec:   opts.TimeSpec,
+		duration:   l.duration,
+	}
+}
+
+// NewLinearRegressionOp creates a new base temporal transform for linear regression functions
+func NewLinearRegressionOp(args []interface{}, optype string) (transform.Params, error) {
+	var (
+		duration float64
+		ok       bool
+	)
+
+	switch optype {
+	case PredictLinearType:
+		if len(args) != 2 {
+			return emptyOp, fmt.Errorf("invalid number of args for %s: %d", PredictLinearType, len(args))
+		}
+
+		duration, ok = args[1].(float64)
+		if !ok {
+			return emptyOp, fmt.Errorf("unable to cast to scalar argument: %v for %s", args[1], PredictLinearType)
+		}
+
+	case DerivType:
+	default:
+		return nil, fmt.Errorf("unknown linear regression type: %s", optype)
+	}
+
+	l := linearRegressionProcessor{
+		duration: duration,
+	}
+
+	return newBaseOp(args, optype, l)
+}
+
+type linearRegressionNode struct {
+	op         baseOp
+	controller *transform.Controller
+	timeSpec   transform.TimeSpec
+	duration   float64
+}
+
+func (l linearRegressionNode) Process(dps ts.Datapoints, alignedTime time.Time) float64 {
+	if dps.Len() < 2 {
+		return math.NaN()
+	}
+
+	var slope, intercept float64
+
+	if l.op.operatorType == PredictLinearType {
+		slope, intercept = linearRegression(dps, alignedTime)
+		return slope*l.duration + intercept
+	}
+
+	slope, _ = linearRegression(dps, time.Time{})
+	return slope
+}
+
+// linearRegression performs a least-square linear regression analysis on the
+// provided datapoints. It returns the slope, and the intercept value at the
+// provided time. The algorithm we use comes from https://en.wikipedia.org/wiki/Simple_linear_regression.
+func linearRegression(dps ts.Datapoints, interceptTime time.Time) (float64, float64) {
+	var (
+		n            float64
+		sumX, sumY   float64
+		sumXY, sumX2 float64
+
+		nonNaNCount int
+	)
+
+	for _, dp := range dps {
+		if math.IsNaN(dp.Value) {
+			continue
+		}
+
+		if nonNaNCount == 0 {
+			if interceptTime.IsZero() {
+				// set interceptTime as timestamp of first non-NaN dp
+				interceptTime = dp.Timestamp
+			}
+		}
+
+		nonNaNCount++
+
+		// convert to milliseconds
+		x := float64(dp.Timestamp.Sub(interceptTime).Nanoseconds()/1000000) / 1e3
```

Comment:
I think this converts it to seconds, not millis. Can you just do `float64(dp.Timestamp.Sub(interceptTime).Seconds())`?